### DISCUSSION TITLE
Boundary Condition between 2 variables

### DISCUSSION THREAD

**Discussion by gj19866** [idaholab/moose#26142] [Q&A General]
Hi there,

Thank you for all the help you have offered so far.

The problem I'm solving requires me to use a complex variable. As Moose doesn't handle complex values, I have split it into real and imaginary parts, such that C = a +i*b.

a and b are variables that are solved by kernels here.

I want to set a boundary condition such that the magnitude of C =1, hence a^2+b^2=1. 
How would i go about doing this?

All the best,
Gillian

**Comment by GiudGiud** [comment]
Hello

You can write a dirichlet BC that involves another variable. 

I would use that to set the value of one variable based on the variable of another

Either that or write a constraint instead

Guillaume

**Reply by gj19866** [reply]
Hi Guillamume,

Thanks, for the help.
Yep, that makes sense.
Is there an example that you would recommend for me to look at to write the BC/ base it on?

All the best,
Gillian

**Reply by gj19866** [reply]
I have implemented this, as in the example page here, (https://mooseframework.inl.gov/getting_started/examples_and_tutorials/examples/ex04_bcs.html).

The solve now won't converge. 
I want to clarify, what happens to the boundary condition on the other variable in this case. Will it also be adjusted? Will it default to a different BC?

As you now have one BC that's active on one variable, but actually it is acting on 2 variables.

Thank you

**Comment by gj19866** [comment]
````
#pragma once

#include "NodalBC.h"

/// Implements a coupled Dirichlet BC where u^2 +Psi_im^2 = Psi_mag on the boundary.
class SuperBC : public NodalBC
{
public:
  SuperBC(const InputParameters & parameters);

  static InputParameters validParams();

protected:
  virtual Real computeQpResidual() override;

private:
  /// Multiplier on the boundary.
  Real _psi_MAG;
  /// reference to a user-specifiable coupled (independent) variable
  const VariableValue & _Psi_Im;
};

```

```

#include "SuperBC.h"

registerMooseObject("SCResApp", SuperBC);

InputParameters
SuperBC::validParams()
{
  InputParameters params = NodalBC::validParams();

  // Specify input parameters that we want users to be able to set:
  params.addParam<Real>("psi_MAG", 0, "Mag of Psi on the boundary");
  params.addRequiredCoupledVar("Psi_Im", "Value on the boundary");
  return params;
}

SuperBC::SuperBC(const InputParameters & parameters)
  : NodalBC(parameters),
    // store the user-specified parameters from the input file...
    _psi_MAG(getParam<Real>("psi_MAG")),        // for the multiplier
    _Psi_Im(coupledValue("Psi_Im")) // for the coupled variable
{
}

Real
SuperBC::computeQpResidual()
{
  // For dirichlet BCS, u=BC at the boundary, so the residual includes _u and the desired BC value:
  return _u[_qp]*_u[_qp] + _Psi_Im[_qp]*_Psi_Im[_qp] - (_psi_MAG);
}
```

**Reply by GiudGiud** [reply]
Hello

I don't see a computeJacobian routine. Missing those terms is not going to help convergence. 

You could try using AD? See the page about automatic differentiation in the website


If not then you should check out our page in troubleshooting failed solves. There's good advice on diagnosing convergence issues

Guillaume

**Reply by gj19866** [reply]
I have given it a go writing it as AD, what do i need to change here to make it AD?


I want to clarify, what happens to the boundary condition on the other variable in this case. Will it also be adjusted? Will it default to a different BC?
As you now have one BC that's active on one variable, but actually it is acting on 2 variables, so will both variables be adjusted as required?

**Reply by gj19866** [reply]
I have updated to using AD, that seems like the better option, as most of my code uses AD anyway.

The file now will not make, but i cant understand why.

```

#pragma once

#include "ADNodalBC.h"

/// Implements a coupled Dirichlet BC where u = alpha * some_var on the boundary.
class SuperBC : public ADNodalBC
{
public:
  SuperBC(const InputParameters & parameters);

  static InputParameters validParams();

protected:
  virtual ADReal computeQpResidual() override;

private:
  /// Multiplier on the boundary.
  Real _psi_MAG;
  /// reference to a user-specifiable coupled (independent) variable
  const ADVariableValue & _Psi_Im;
};

```

.C

```

#include "SuperBC.h"

registerMooseObject("SCResApp", SuperBC);

InputParameters
SuperBC::validParams()
{
  InputParameters params = ADNodalBC::validParams();

  // Specify input parameters that we want users to be able to set:
  params.addParam<Real>("psi_MAG", 0, "Mag of Psi on the boundary");
  params.addRequiredCoupledVar("Psi_Im", "Value on the boundary");
  return params;
}

SuperBC::SuperBC(const InputParameters & parameters)
  : ADNodalBC(parameters),
    // store the user-specified parameters from the input file...
    _psi_MAG(getParam<Real>("psi_MAG")),        // for the multiplier
    _Psi_Im(adCoupledValue("Psi_Im")) // for the coupled variable
{
}

ADReal
SuperBC::computeQpResidual()
{
  // For dirichlet BCS, u=BC at the boundary, so the residual includes _u and the desired BC value:
  return _u[_qp]*_u[_qp] + _Psi_Im[_qp]*_Psi_Im[_qp] - (_psi_MAG);

}

```

It is giving this error:

```

Checking if header needs updating: /home/lstein/projects/moose/framework/include/base/MooseRevision.h...
Creating Unity /home/lstein/projects/sc_res/build/unity_src/BCs_Unity.C
Compiling C++ (in opt mode) /home/lstein/projects/sc_res/build/unity_src/BCs_Unity.C...
In file included from /home/lstein/projects/sc_res/build/unity_src/BCs_Unity.C:2:
/home/lstein/projects/sc_res/src/BCs/SuperBC.C: In member function 'virtual ADReal SuperBC::computeQpResidual()':
/home/lstein/projects/sc_res/src/BCs/SuperBC.C:29:10: error: no match for 'operator[]' (operand types are 'const type' {aka 'const MetaPhysicL::DualNumber<double, MetaPhysicL::SemiDynamicSparseNumberArray<double, long unsigned int, MetaPhysicL::NWrapper<53> >, true>'} and 'const unsigned int')
   28 |   return _u[_qp]*_u[_qp] + _Psi_Im[_qp]*_Psi_Im[_qp] - (_psi_MAG);
      |                    ^
make: *** [/home/lstein/projects/moose/framework/build.mk:149: /home/lstein/projects/sc_res/build/unity_src/BCs_Unity.x86_64-conda-linux-gnu.opt.lo] Error 1


```

**Reply by GiudGiud** [reply]
Weird. I dont see why that would be wrong. Can you try this?
```
auto u_value = _u[_qp];
return u_value * u_value + _Psi_Im[_qp]*_Psi_Im[_qp] - (_psi_MAG);
```

**Reply by gj19866** [reply]
Still not working:

````
/home/lstein/projects/sc_res/src/BCs/SuperBC.C: In member function 'virtual ADReal SuperBC::computeQpResidual()':
/home/lstein/projects/sc_res/src/BCs/SuperBC.C:28:20: error: no match for 'operator[]' (operand types are 'const type' {aka 'const MetaPhysicL::DualNumber<double, MetaPhysicL::SemiDynamicSparseNumberArray<double, long unsigned int, MetaPhysicL::NWrapper<53> >, true>'} and 'const unsigned int')
   28 |   auto u_value = _u[_qp];
      |                    ^
make: *** [/home/lstein/projects/moose/framework/build.mk:149: /home/lstein/projects/sc_res/build/unity_src/BCs_Unity.x86_64-conda-linux-gnu.opt.lo] Error 1

```

**Reply by GiudGiud** [reply]
ah ok at least this is consistent.

I think because there is only 1 value to _u on a node (there is no qp loop on a node) you do NOT need to index.

so just `_u` not `_u[_qp]`

**Reply by gj19866** [reply]
thank you, 
that has fixed the making of it.
The solve still isn't converging. 

I want to clarify, what happens to the boundary condition on the other variable in this case. Will it also be adjusted? Will it default to a different BC?
As you now have one BC that's active on one variable, but actually it is acting on 2 variables, so will both variables be adjusted as required?

**Reply by GiudGiud** [reply]
It wont get 'adjusted' but it should converge to a configuration that meets the equation you imposed with the nodal BC. What happens to the BC also depends a lot on how you wrote the equations in the kernels.

Can you set `preset = false` to the BC?
btw what variable types (finite element type) are the two variables?
This is the troubleshooting advice page btw:
https://mooseframework.inl.gov/moose/application_usage/failed_solves.html

**Reply by gj19866** [reply]
The variable types are both first Lagrange.
It is now all working.
Thank you very much Guillamume!

**Reply by gj19866** [reply]
Looking closer at my results, it appears now that the variables _u and Psi_Im are not able to change from the initial condition set for them on the boundary. 
I had hopped that they would still change, just with the additional constraint u^2 + psi_im^2 = 1 on the boundary.
Any ideas?

**Reply by GiudGiud** [reply]
I was wondering if a nodalBC made less sense than a DirichletBC for these. Do you mind looking into using a DirichletBC?

**Reply by gj19866** [reply]
Yep, I will look at using a DirichletBC.

Why would that make more sense/ let the value change?
Thanks

**Reply by gj19866** [reply]
Hi GuidGuid, would it be possible to get a bit more help with this?

I have tried writing it as a DirichletBC as below, but it doesn't seem to be working either.

Do you have any ideas for me?
Thank you for all the help so far!

```

#pragma once

#include "ADDirichletBCBase.h"

/// Implements a coupled Dirichlet BC where u = alpha * some_var on the boundary.
class SuperBC2 : public ADDirichletBCBase
{
public:
  SuperBC2(const InputParameters & parameters);

  static InputParameters validParams();

protected:
  virtual ADReal computeQpValue() override;

private:
  /// Multiplier on the boundary.
  Real _psi_MAG;
  /// reference to a user-specifiable coupled (independent) variable
  const ADVariableValue & _Psi_Im;
};


```


````

#include "SuperBC2.h"
#include <cmath>


registerMooseObject("SCResApp", SuperBC2);

InputParameters
SuperBC2::validParams()
{
  InputParameters params = ADDirichletBCBase::validParams();

  // Specify input parameters that we want users to be able to set:
  params.addParam<Real>("psi_MAG", 0, "Mag of Psi on the boundary");
  params.addRequiredCoupledVar("Psi_Im", "Value on the boundary");
  return params;
}

SuperBC2::SuperBC2(const InputParameters & parameters)
  : ADDirichletBCBase(parameters),
    // store the user-specified parameters from the input file...
    _psi_MAG(getParam<Real>("psi_MAG")),        // for the multiplier
    _Psi_Im(adCoupledValue("Psi_Im")) // for the coupled variable
{
}

ADReal
SuperBC2::computeQpValue()
{
  // For dirichlet BCS, u=BC at the boundary, so the residual includes _u and the desired BC value:
return std::sqrt((_psi_MAG) - (_Psi_Im*_Psi_Im));

}

```

**Reply by GiudGiud** [reply]
have you tried the troubleshooting instructions yet?
https://mooseframework.inl.gov/moose/application_usage/failed_solves.html

**Reply by gj19866** [reply]
The solve failing isnt the issue. The solve converges fine, however the values at the boundary do not change from their initial value as the rest of the solve equilibriates.

![Screenshot 2023-12-01 123256](https://github.com/idaholab/moose/assets/141067346/b0495f1a-2814-4172-9cdc-750ce3d5cb59)

I want the values to equilibriate, and u and Psi_im to change, whilst being held with the condition that u^2+psi_im^2=1.

Does that make sense?

Thank you 
Gillian

**Reply by GiudGiud** [reply]
So the boundary condition is obeyed, but the values do not equilibriate?
Are both variables not updating? u and phi_im ?
Are the variables supposed to diffuse / advect / ??? from the boundary into the volume?

**Reply by gj19866** [reply]
Yes, that is correct, that values do not equilibrate to the bulk value.
Both variables are not updating. Both u and psi_im.

The values do not diffuse or advect as such, as this is modelling a superconductor, but the values should adjust such that all the values are the same, which is not happening. 

All the best,
Gillian Lloyd

**Reply by lindsayad** [reply]
> The values do not diffuse or advect as such, as this is modelling a superconductor, but the values should adjust such that all the values are the same,

Why should they become the same if there is not some spatial migration effect to make it so? What are the equations in the volume?

**Reply by gj19866** [reply]
I see your point. 
On reflection, there are diffusion terms but bundled in with other terms, as below.
Apologies for the confusion.

The equations being solved are as follows:

$$
0=-u \sqrt{1+\gamma^2\left(\psi_R^2+\psi_I^2\right)}\left(\frac{\partial}{\partial t} \psi_R-\varphi \psi_I\right)+\nabla^2 \psi_R-\gamma^2 \psi_I \nabla^2 \varphi+\left(1-\psi_R^2-\psi_I^2\right) \psi_R
$$

$$
0=-u \sqrt{1+\gamma^2\left(\psi_R^2+\psi_I^2\right)}\left(\frac{\partial}{\partial t} \psi_I+\varphi \psi_R\right)+\nabla^2 \psi_I+\gamma^2 \psi_R \nabla^2 \varphi+\left(1-\psi_R^2-\psi_I^2\right) \psi_I
$$

$$
0=\nabla^2 \varphi-\psi_R \nabla^2 \psi_I+\psi_I \nabla^2 \psi_R
$$

**Reply by lindsayad** [reply]
You mention `u` a lot in previous posts. Is `u` equivalent to phi in the comment immediately above?

**Reply by gj19866** [reply]
I was referencing `u` as that was the form in the kernel, but `u` is equivalent to `psi_Re` in this set of equations. 

So essentially the BC I'm looking for for this set of equations is that $\psi_R^2+\psi_I^2 = 1$.

Is that okay? Sorry if I'm not being clear.

**Reply by lindsayad** [reply]
In your code above you defaulted the magnitude of Psi to 0. That seems like a poor default. Are you changing it to 1 in your input file? If the magnitude is 0, then the only solution is for both u and `psi_Im` to be 0

**Reply by gj19866** [reply]
The value is set to 1 in the input file. 
The idea of it being defaulted to 0 is that if not specified it will default to the solution where u=psi_im=0. This is another case that I am looking at, which of course is the same as the normally dirichlet BC = 0.
I will change the default value to 1 to prevent confusion. 

The issues I am having with the variables not changing are in the case where $u^2 + \psi_I^2 = \psi_{MAG}^2 = 1$ however.

**Reply by lindsayad** [reply]
So it seems like your boundary condition is not being violated so I think we should move our thinking beyond the BC you implemented. You have created one boundary condition and have applied it to the `u` degrees of freedom. What boundary condition are you applying/would you like to apply for the `psi_Im` degrees of freedom?

**Reply by gj19866** [reply]
Yes, I agree. 

This is where I am confused, and I belive the issue with my implementation lies. 
I am not currently applying a BC aside from this on `psi_Im`. 
The only constraint I want on the values of `u `, and `psi_Im` here is that $u^2 + \psi_Im^2 = 1$, and hence I thought that if I added another BC then this would be overconstraining the problem. Is this correct?

Thank you for helping me.

**Reply by lindsayad** [reply]
So in your equation set, you have Laplacian terms, so you have diffusion as you pointed out. I assume that you are integrating these diffusion terms by parts? If you are, then by not not applying a boundary condition for `psi_Im` you are actually applying one ... the natural boundary condition. The natural condition weakly sets your surface terms arising from integration by parts to 0. So for a pure diffusion equation, the natural condition is that the diffusive flux is zero at the boundary, e.g. for a 1D problem dfoo/dx = 0.

**Reply by gj19866** [reply]
Ah yes, that makes sense to me. 
And yes I do integrate by parts. 
I have attached the weak forms below.

How would you suggest I alter the surface terms in that case then?
Thank you for helping, I have not had to write my own BC before in MOOSE?

The BC i am trying to write is that $\psi_R^2+\psi_I^2=1$
![Screenshot 2023-12-05 202726](https://github.com/idaholab/moose/assets/141067346/7a7a515c-77af-44e7-888d-5a4280ecdd3e)

**Reply by lindsayad** [reply]
What are the physics you're trying to solve here? Electromagnetics? Wave propagation?

**Reply by YaqiWang** [reply]
You have two variables \psi_I and \psi_R but only have one test function \mu? And what is that \varphi? Can you give the PDE and BC?

**Reply by gj19866** [reply]
I am solving for $\psi_I$, $\psi_R$, and $\varphi$. 

The equations being solved are:

$$
0=-u \sqrt{1+\gamma^2\left(\psi_R^2+\psi_I^2\right)}\left(\frac{\partial}{\partial t} \psi_R-\varphi \psi_I\right)+\nabla^2 \psi_R-\gamma^2 \psi_I \nabla^2 \varphi+\left(1-\psi_R^2-\psi_I^2\right) \psi_R
$$

$$
0=-u \sqrt{1+\gamma^2\left(\psi_R^2+\psi_I^2\right)}\left(\frac{\partial}{\partial t} \psi_I+\varphi \psi_R\right)+\nabla^2 \psi_I+\gamma^2 \psi_R \nabla^2 \varphi+\left(1-\psi_R^2-\psi_I^2\right) \psi_I
$$

$$
0=\nabla^2 \varphi-\psi_R \nabla^2 \psi_I+\psi_I \nabla^2 \psi_R
$$

And these are currently solved without any problems. 

What I want is to implement a BC where $\psi_R^2+\psi_I^2=1$. This is essentially that I want to constrain the magnitude of complex order parameter $\psi$, but not the phase of $\psi$. $\psi = \psi_R + i \psi_I$.

**Reply by YaqiWang** [reply]
I see. Small correction to your weak form: it should differentiate $\mu$ into $\mu_R$, $\mu_I$ and $\mu$ for three equations. I am not sure whether you can use a Lagrange multiplier $\lambda$ which is defined on the boundary, then you can have $(\lambda^*, \psi_R^2 + \psi_I^2 - 1)$ to enforce the BC. But typically $\lambda$ should have a physical meaning and should be used in the current three equations, which I have no idea on.

**Comment by gj19866** [comment]
> What are the physics you're trying to solve here? Electromagnetics? Wave propagation?

I am modelling the behaviour of superconductors as they move from a superconducting to a normal resistive state by solving the Time-Dependent Ginzburg-Landau Equations for a dirty superconductor. https://link.springer.com/article/10.1007/BF00683522

The model is working well currently and shows the behaviour expected for boundary conditions where $|\psi|=0$. But I would like to model superconducting boundary conditions, where the magnitude of the complex order parameter $|\psi|=1$. Following your advice, I handle the real and imaginary parts of the order parameter separately, so I am looking to implement that $\psi_{Re}^2+\psi_{Im}^2=1$.

**Reply by GiudGiud** [reply]
@cticenhour @aaelmeli I think you have worked on similar problems. Do you have advice on doing this?

**Reply by lindsayad** [reply]
Do you use the same BC for |psi| = 0?

**Reply by aaelmeli** [reply]
Hi

You may want to implement something like `CoupledVarNeumannBC` that is customized to your case. I think the current implementation does not help you but it may be a good start to get to what you want. The following is an example of how I used `CoupledVarNeumannBC` to couple the real and imaginary variables scaled with a coefficient on some boundary. Also, consider using a direct solver for your verifications so that you make sure that the Jacobian is computed accurately. 

```
    type = CoupledVarNeumannBC
    variable = uyi #imaginar variable
    boundary = '3'
    v = uyr #real variable
    coef = '${fparse -rho*omega*C_sr}' #some scaling factor which is a function of the shear wave speed, frequency and density"
```

**Reply by gj19866** [reply]
Hi @aaelmeli,

I have had a look at the `CoupledVarNeumannBC`, and it looks very similar to what I have been doing.

I notice that the difference is that `CoupledVarNeumannBC` uses an `IntegratedBC`. Is this something I should try instead of the `NodalBC`? Does that make sense in this case?

Just to clarify, when you use `CoupledVarNeumannBC` are the values of both `uyi` and `uyr` changed/adjusted on the boundary in this case, or does one of these values stay fixed?

**Reply by gj19866** [reply]
For reference this is my BC:

```

#include "SuperBC.h"

registerMooseObject("SCResApp", SuperBC);

InputParameters
SuperBC::validParams()
{
  InputParameters params = ADNodalBC::validParams();

  // Specify input parameters that we want users to be able to set:
  params.addParam<Real>("psi_MAG", 1, "Mag of Psi on the boundary");
  params.addRequiredCoupledVar("Psi_Im", "Value on the boundary");
  return params;
}

SuperBC::SuperBC(const InputParameters & parameters)
  : ADNodalBC(parameters),
    // store the user-specified parameters from the input file...
    _psi_MAG(getParam<Real>("psi_MAG")),        // for the multiplier
    _Psi_Im(adCoupledValue("Psi_Im")) // for the coupled variable
{
}

ADReal
SuperBC::computeQpResidual()
{
  // For dirichlet BCS, u=BC at the boundary, so the residual includes _u and the desired BC value:
  auto u_value = _u;
//   return _u[_qp]*_u[_qp] + _Psi_Im[_qp]*_Psi_Im[_qp] - (_psi_MAG);
return u_value*u_value  + _Psi_Im[_qp]*_Psi_Im[_qp] - (_psi_MAG);

}

```


```

#pragma once

#include "ADNodalBC.h"


class SuperBC : public ADNodalBC
{
public:
  SuperBC(const InputParameters & parameters);

  static InputParameters validParams();

protected:
  virtual ADReal computeQpResidual() override;

  /// Multiplier on the boundary.
  Real _psi_MAG;
  /// reference to a user-specifiable coupled (independent) variable
  const ADVariableValue & _Psi_Im;
};

```

**Reply by gj19866** [reply]
It would make sense to me if it isn't possible to write a BC in this way, as I can't imagine wanting a BC to act on, and adjust, 2 variables is standard. 

Thanks for all the help so far everyone.

**Reply by lindsayad** [reply]
Writing boundary conditions for degrees of freedom that are functions of other degrees of freedom is very common, so in that sense what you are doing is "standard"

**Reply by gj19866** [reply]
Thats reassuring. 
If the issue is to do with the natural condition being applied that the flux across the boundary is zero, how would I stop enforcing that, without setting it to a value as such?

**Reply by lindsayad** [reply]
You can write a "do-nothing" boundary condition. So essentially you can write a BC that is your surface term coming from integration by parts

**Comment by gj19866** [comment]
> Do you use the same BC for |psi| = 0?

To enforce |psi|=0 I have just been using a standard Dirichlet BC, as psi_R=psi_I=0 in this case, so I presumed it would be better practice (and more computationally efficient?) to use the Dirichlet BC in that case.

**Reply by lynnmunday** [reply]
We do something like this in tensor mechanics with:  https://mooseframework.inl.gov/source/bcs/PenaltyInclinedNoDisplacementBC.html
where we want to enforce a zero displacement on the normal component of a surface so that it remains planar.  To do that we need to "tie" together the displacement components on that surface.  Displacement is a vector so each component is its own degree of freedom.  This enforces a dirichlet boundary condition weakly.  Maybe @jiangwen84 or @hugary1995 or @bwspenc would know if the penalty bc could enforce the bc you want.
